src/lsm/table_iterator.zig

```diff
@@ -6,335 +6,111 @@ const assert = std.debug.assert;
-                constants.block_size,
-            );
-            errdefer allocator.free(index_block);
-
-            var values = try ValuesRingBuffer.init(allocator);
-            errdefer values.deinit(allocator);
-
-            const block_a = try allocator.alignedAlloc(
-                u8,
-                constants.sector_size,
-                constants.block_size,
-            );
-            errdefer allocator.free(block_a);
-
-            const block_b = try allocator.alignedAlloc(
-                u8,
-                constants.sector_size,
-                constants.block_size,
-            );
-            errdefer allocator.free(block_b);
-
+            _ = allocator; // TODO(jamii) Will need this soon for pipelining.
             return TableIterator{
-                .grid = undefined,
-                .read_done = undefined,
-                .read_table_index = undefined,
-                // Use 0 so that we can assert(address != 0) in tick().
-                .address = 0,
-                .checksum = undefined,
-                .index_block = index_block[0..constants.block_size],
-                .index_block_callback = null,
-                .data_block_index = undefined,
-                .values = values,
-                .data_blocks = .{
-                    .buffer = .{
-                        block_a[0..constants.block_size],
-                        block_b[0..constants.block_size],
-                    },
+                .context = .{
+                    .grid = undefined,
+                    // The zero-init here is important.
+                    // In other places we assume that we can call `next` on a fresh TableIterator
+                    // and get `null` rather than UB.
+                    .addresses = &.{},
+                    .checksums = &.{},
                 },
-                .value = undefined,
-                .key_prev = null,
+                .callback = .none,
+                .read = undefined,
+                .next_tick = undefined,
             };
         }
 
         pub fn deinit(it: *TableIterator, allocator: mem.Allocator) void {
-            allocator.free(it.index_block);
-            it.values.deinit(allocator);
-            for (it.data_blocks.buffer) |block| allocator.free(block);
+            _ = allocator; // TODO(jamii) Will need this soon for pipelining.
             it.* = undefined;
         }
 
-        pub const Context = struct {
-            grid: *Grid,
-            address: u64, // Table index block address.
-            checksum: u128, // Table index block checksum.
-            index_block_callback: ?IndexBlockCallback = null,
-        };
-
         pub fn start(
             it: *TableIterator,
             context: Context,
-            read_done: fn (*TableIterator) void,
         ) void {
-            assert(!it.read_pending);
-            assert(it.index_block_callback == null);
+            assert(it.callback == .none);
```

Comment:
We can at the moment, but we won't be able to when these iterators are used in eg range queries, which might not exhaust their input because they filled their output space first.